### ISSUE TITLE
Deep-linking capability on Accordion and Tabs components

### ISSUE THREAD

**Issue by rajeevyadav2** [adobe/aem-core-wcm-components#942]
## Feature Request

**Is your feature request related to a problem? Please describe.**
Content team should be able to anchor URLs to specific item/panel on these components.

**Describe the solution you'd like**
Each Item/Panel in accordion should have unique id.
Javascript should check the anchor and open the specific panel/item in the component and scroll the window to bring user on the anchored panel.

**Are there alternatives?**
No

**Documentation**
If you can, explain how users will be able to use this and possibly write out a version of the docs.
Maybe a screenshot or design?

**Comment by gabrielwalt** [comment]
+1 this is a feature we've also been thinking about for long.

We did a first step towards this feature with #920 ("Configurable IDs for all components"). Now a second step would be to adjust the JS client library of the accordion and tab components to automatically open the corresponding panel when that panel's ID is in the URL hash.

If you'd like to tackle this, @rajeevyadav2, your contribution is welcome!

**Comment by rajeevyadav2** [comment]
@gabrielwalt : Thank you for the information. Do I need to open RTC first before starting working on it or else I should fork the repository and start working on changes?

**Comment by rajramo61** [comment]
@gabrielwalt : I will be working on this issue with @rajeevyadav2 .

**Comment by rajeevyadav2** [comment]
@gabrielwalt : I was looking into ID implementation in tab and accordion component in development branch and see issues with the approach taken.

As per the approach, each item id is based on index number of the item. Now lets say some item is at 3rd index and marketing team configures campaign urls in various channels to point anchor to 3rd item. Post this if any reordering is done in the item, campaign urls will still be pointing to 3rd item which may not be correct. I feel the Id should always remain same after it has been generated and should not change once it moves from one location to another. This will ensure that anchor always points to correct content.

Do let me know your thoughts on this.

**Comment by bpauli** [comment]
@rajeevyadav2 Indeed, the index of the accordion can always change. Do you think it would be sufficient if we use the `ListItem#getName()` as a suffix for the accordion item id?

**Comment by jckautzmann** [comment]
@bpauli Ideally all links (all linkable items) should have a unique ID, similar to the component ID. It will be needed e.g to identify a link in the data layer.

**Comment by bpauli** [comment]
@jckautzmann If the id is a combination of accordion id and ListItem#getName (aka Resource#getName = last segment of the JCR path) it will be unique `${accordion.id}-panel-${item.name}` as a child resource of a container has always a unique name. The only issue I see, is that it might be not URL decoded

**Comment by rajeevyadav2** [comment]
> @rajeevyadav2 Indeed, the index of the accordion can always change. Do you think it would be sufficient if we use the `ListItem#getName()` as a suffix for the accordion item id?

@bpauli : Yes. Item name should be good.

**Comment by jckautzmann** [comment]
@bpauli Another option is: `accordion-panel-${item.id}`, id being computed from the hash of the item path (same logic as for `Component#getId`)

**Comment by bpauli** [comment]
@jckautzmann I was thinking the same but this would require some more work.

We have to extend the API of the ListItem. Each implementation of the ListItem must implement this method. There are currently 11 classes which are implementing this interface but the id makes only sense for one of them. We should not extend form AbstractComponent so we have to extract the id generation stuff to another location. It's not possible for all components to edit the id attribute (e.g. Layout Container which doesn't have an edit dialog).

Using the name of the JCR node would require much less work and would fit to the requirements.

**Comment by gabrielwalt** [comment]
@rajeevyadav2 and @rajramo61, sure go ahead with creating a PR, we're looking forward to it very much, it sounds like a great capability!

**Comment by rajramo61** [comment]
@gabrielwalt and @bpauli 
I worked with @rajeevyadav2 and have added the PR listed above. Please review.

**Comment by gabrielwalt** [comment]
Very nice, thank you! We'll review your PR in the coming days.

**Comment by rajramo61** [comment]
@gabrielwalt , @bpauli and @richardhand created the PR for Tabs component. Last PR covered Accordion component. Here is the PR link
https://github.com/adobe/aem-core-wcm-components/pull/960

cc @rajeevyadav2

**Comment by rajramo61** [comment]
@gabrielwalt , @rajeevyadav2, @bpauli and @richardhand created the PR for Tabs component. Last PR covered Accordion component.

**Comment by jckautzmann** [comment]
See also issue #953.

**Comment by jckautzmann** [comment]
@rajramo61 @rajeevyadav2
As the implementation for item ID has changed, I created a new pull request [0] that is based on the code provided through the two existing pull requests [1][2].

[0] https://github.com/adobe/aem-core-wcm-components/pull/1011
[1] https://github.com/adobe/aem-core-wcm-components/pull/946
[2] https://github.com/adobe/aem-core-wcm-components/pull/960

Does it match your expectations?
